execution-engine/common/src/contract_api/mod.rs

```
@@ -387,6 +387,19 @@ pub fn create_purse() -> PurseId {
     }
 }
 
+pub fn main_purse() -> PurseId {
+    // TODO: this could be more efficient, bringing the entire account
+    // object across the host/wasm boundary only to use 32 bytes of
+    // its data is pretty bad. A native FFI (as opposed to a library
+    // API) would get around this problem. However, this solution
+    // works for the time being.
+    // https://casperlabs.atlassian.net/browse/EE-439
+    let account_pk = get_caller();
+    let key = Key::Account(account_pk.value());
+    let account: Account = read_untyped(&key).unwrap().try_into().unwrap();
+    account.purse_id()
+}
```

why not just add a get_caller_purseid method on  the host?

---

Yes, as discussed on Slack: https://casperlabs.atlassian.net/browse/EE-439

---

Right. That's what I meant by a "native FFI" in the TODO above (with associated ticket). I just didn't want to do it right now because adding a whole new FFI is much more work than just adding a new library API.

---

This will fail with the `InvalidAccess` error in a subcall. That's because `base_key` in the subcall will be its [contract's] address in the GlobalState and `read_untyped(Key::Account(address))` will compare whether `Key::Account(address) == base_key`.

---

That sounds like a feature as opposed to a bug since contracts shouldn't be able to access the purse of caller.

---

I don't think contracts need any API like this because they don't have a default purse. User purses a contract interacts with would be passed in as arguments and internal purses it would manage itself the same way it would any internal state (using its named keys or local state).

---

But that method is in the API. What will happen if I call `main_purse()` in the contract context?

---

Didn't you say in your first comment that nothing will get returned because it'll throw an error when it tries to read the Account because that key is only readable by the account itself (i.e. in session code only)?

It's a good question though when thinking about what we want the proper FFI to look like. We could keep the same behaviour (let it just error out), or have it return an option. I'm kind of leaning towards the former because it will always succeed in session code and always fail in sub-calls so the option is not really helpful for control flow.

---

In my first comment I meant that will happen with _current_ implementation - as it reads the whole account and then extracts the purse.

I imagine that next iteration will provide a `main_purse` API method that reads current account's purse. Unless in the subcall.

---

>  I'm kind of leaning towards the former because it will always succeed in session code and always fail in sub-calls so the option is not really helpful for control flow.

So it will fail, with an error, that stops the execution?

---

Yes, that's my initial feeling. But I'm open to counter arguments or other suggestions.